SIPS/sip-20.md

```diff
@@ -1,191 +1,301 @@
+
+### Top-level architecture
+
+This architecture defers WebSocket management to the client, allowing Snaps to open, close, and send messages over WebSockets. The client will handle the underlying WebSocket connections and notify the Snap of any events that occur on those connections.
+
+The client has multiple responsibilities in this architecture:
+
+- All WebSocket connections opened by Snaps MUST be monitored by the client. Each connection is identified by a unique identifier, which is returned to the Snap when the connection is opened.
+
+- Any incoming message from WebSocket connections MUST be handled by the client and forwarded to the appropriate Snap through `onWebSocketEvent`. 
+
+- When the Snap requests to open a WebSocket connection via `snap_openWebSocket`, the client MUST ensure that the connection is established and return a unique identifier for that connection.
+
+- Any open connection MUST be kept alive by the client until the Snap explicitly closes it or the connection is lost, even when the Snap is terminated.
+
+- When the Snap requests to close a WebSocket connection via `snap_closeWebSocket`, the client MUST close the connection and remove it from its list of open connections.
+
+- Any errors that occur when the WebSocket connection is open (e.g., connection failure, message send failure) MUST be handled by the client and reported to the Snap via `onWebSocketEvent`.
 
-### Snap manifest
 
-This SIP introduces a new permission named `endowment:external-data`.
+### Snap Manifest
+
+This SIP doesn't introduce any new permissions, but rather extends `endowment:network-access` with new capabilities.
+
+### RPC Methods
+
+#### `snap_openWebSocket`
+
+This method allows a Snap to open a WebSocket connection to an external service. The method takes a URL as a parameter and returns a unique identifier for the connection.
 
 ```typescript
-/**
- * A WebSocket connection as specified in the Snap manifest.
- *
- * @property type - The type of connection. Currently, only WebSockets are
- * supported.
- * @property url - The URL to connect to. This must be a valid WebSocket URL,
- * starting with `wss://`. URLs starting with `ws://` are not allowed.
- * @property dataType - The type of data expected from the WebSocket. If this
- * field is omitted, text is assumed.
- */
-type WebSocketConnection = {
-  type: 'websocket';
+
+type OpenWebSocketParams = {
   url: string;
-  dataType?: 'text' | 'binary';
+  protocols?: string[];
 };
+```
+The RPC method takes one parameter:
+
+- `url` - The URL of the WebSocket service to connect to. 
+  - The URL MUST be a valid WebSocket URL, starting with `wss://`. URLs starting with `ws://` are not allowed.
+  - Only one WebSocket connection can be opened per URL at a time. If a Snap tries to open a new connection to the same URL while an existing connection is still open, the client MUST throw an error.
+- `protocols` - An optional array of subprotocols to use for the WebSocket connection.
```

Comment:
What is the usage for this field?

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket#protocols

Reply:
Just a note: if, for some reason, the Snap needs to use two sub-protocols, it will need to open two connections:

> Note however that only one sub-protocol can be selected per connection.